Lesson 1 8

Victoria240 opened the pull request:
Data Fetching, Re-Fetching, and Memoized Handlers

Commit:
fetched data from Airtable's API with useEffect

Commit:
added a POST feature to add todos to the Airtable API

Commit:
Merge remote-tracking branch 'origin/lesson_1_7' into lesson_1_8

Commit:
task re-do

Commit:
task re-do/made POST request

Commit:
task re-do

Commit:
task re-do/made a POST Request

Commit:
added React Router

dargaCode commented on src/App.js:
```diff
@@ -1,63 +1,142 @@
+  }
+
+
+  // Use useEffect to simulate loading state and fetch data from localStorage with a 2 seconds delay
+  useEffect(() => {
+    setTimeout(() => {
+      const savedTodoList = JSON.parse(localStorage.getItem("savedTodoList")) || [];
+      setTodoList(savedTodoList);
+      setIsLoading(false);
+    }, 2000);
+  }, []); // Empty dependency array, so this effect runs once after initial render
+
```
Now that you're fetching real data from Airtable, you no longer need to save nor load anything from `LocalStorage`! You can delete this effect entirely =D

dargaCode commented on src/App.js:
```diff
@@ -1,63 +1,142 @@
+  }, []);
+
+
+  // Use useEffect to manage localStorage and persist todoList
+  useEffect(() => {
+    if (!isLoading) {
+      localStorage.setItem('savedTodoList', JSON.stringify(todoList)); // Save the todoList in localStorage when it changes
     }
+  }, [todoList, isLoading]);
```
Now that you're fetching real data from Airtable, you no longer need to save nor load anything from LocalStorage! You can delete this effect entirely =D

Victoria240 replied:
Thank you for the feedback! I hope it's fine for me to leave it in here and merge this repo as it is, because I want to avoid any conflicts. I'll delete it in my next submission

dargaCode commented on src/App.js:
```diff
@@ -1,63 +1,142 @@
 
-      </>
-    );
+      const data = await response.json();
+
+
+      // Update the todoList state by adding the new todo
+      setTodoList([...todoList, { id: data.id, title: newTodo.title }]);
```
There's a small hidden bug here, which is that `data` is still an object with an array, and that array contains the actual todo that was created: 

<img width="761" alt="Screen Shot 2023-08-24 at 12 09 06 AM" src="https://github.com/Victoria240/react-todo/assets/12771651/447651b1-d24d-4a23-af9d-a793307462d0">

dargaCode replied:
so to actually get the id and title confirmed by the server, you'll have to "unwrap" the layers of data around the todo: 

```suggestion
      const data = await response.json();

      const confirmedTodo = data.records[0];

      // Update the todoList state by adding the new todo
      setTodoList([...todoList, { id: confirmedTodo.id, title: confirmedTodo.fields.title }]);
```

dargaCode replied:
I also want to point call attention to changing `newTodo.title` to `confirmedTodo.fields.title` - 

Because the backend is supposed to be the "source of truth" for all of our data, we should trust the title returned from Airtable over the title that we sent to Airtable, just in case they somehow become different. It won't make a difference in this case, but it's good to keep in mind for the future!

dargaCode reviewed (approved):
Great refactor, this is looking awesome now! 💪💪💪

I pointed out a couple small changes to make before you merge, but you're good to go after that!

Victoria240 committed:
Changing newTodo.title to confirmedTodo.fields.title -

The backend is supposed to be the "source of truth" for all of our data,  so we should trust the title returned from Airtable over the title that we sent to Airtable, just in case they somehow become different.

Co-authored-by: Darga Darga <github@dargacode.com>

Commit:
minor changes

Commit:
changing newTodo.title to confirmedTodo.fields.title - this is because the Airtable/backend data is the source of truth(when there's a change) and not the todo app
